src/core/lib/channel/metrics.h

```diff
@@ -0,0 +1,225 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GRPC_SRC_CORE_LIB_CHANNEL_METRICS_H
+#define GRPC_SRC_CORE_LIB_CHANNEL_METRICS_H
+
+#include <grpc/support/port_platform.h>
+
+#include <atomic>
+#include <cstdint>
+#include <memory>
+#include <vector>
+
+#include "absl/strings/string_view.h"
+
+#include <grpc/support/log.h>
+
+#include "src/core/lib/channel/channel_args.h"
+#include "src/core/lib/gprpp/sync.h"
+
+namespace grpc_core {
+
+// Static only.
+class GlobalInstrumentsRegistry {
+ public:
+  enum class Type {
+    kUndefined,
+    kUInt64,
+    kDouble,
+  };
+  struct GlobalCounterDescriptor {
+    Type type;
+    absl::string_view name;
+    absl::string_view description;
+    absl::string_view unit;
+    std::vector<absl::string_view> label_keys;
+    std::vector<absl::string_view> optional_label_keys;
+  };
+  struct GlobalHistogramDescriptor {
+    Type type;
+    absl::string_view name;
+    absl::string_view description;
+    absl::string_view unit;
+    std::vector<absl::string_view> label_keys;
+    std::vector<absl::string_view> optional_label_keys;
+  };
+  struct GlobalHandle {
+    uint32_t index;
+  };
+  struct GlobalUInt64CounterHandle : GlobalHandle {};
+  struct GlobalDoubleCounterHandle : GlobalHandle {};
+  struct GlobalUInt64HistogramHandle : GlobalHandle {};
+  struct GlobalDoubleHistogramHandle : GlobalHandle {};
+
+  // Creates instrument in the GlobalInstrumentsRegistry.
+  static GlobalUInt64CounterHandle RegisterUInt64Counter(
+      absl::string_view name, absl::string_view description,
+      absl::string_view unit, std::vector<absl::string_view> label_keys,
+      std::vector<absl::string_view> optional_label_keys);
+  static GlobalDoubleCounterHandle RegisterDoubleCounter(
+      absl::string_view name, absl::string_view description,
+      absl::string_view unit, std::vector<absl::string_view> label_keys,
+      std::vector<absl::string_view> optional_label_keys);
+  static GlobalUInt64HistogramHandle RegisterUInt64Histogram(
+      absl::string_view name, absl::string_view description,
+      absl::string_view unit, std::vector<absl::string_view> label_keys,
+      std::vector<absl::string_view> optional_label_keys);
+  static GlobalDoubleHistogramHandle RegisterDoubleHistogram(
+      absl::string_view name, absl::string_view description,
+      absl::string_view unit, std::vector<absl::string_view> label_keys,
+      std::vector<absl::string_view> optional_label_keys);
+
+  static const GlobalCounterDescriptor& GetCounterDescriptor(
```

Comment:
I don't see the point of the `GetCounterDescriptor()` or `GetHistogramDescriptor()` methods. We could just use the `counters()` or `histograms()` methods and use the handle as an index directly?

Reply:
If you are thinking of this as just a convenience method, I'm fine with it

Reply:
Removed. I combined `counters()` and `histograms()` into a single `instruments()` method, PTAL.